### PR TITLE
Add Clang/LLVM to Toolchains

### PR DESCRIPTION
**Pull request by lenary**
This PR adds support for building Clang/LLVM to the built toolchains.

- The existing GCC toolchain tarballs remain largely unmodified (especially as far as Ibex is concerned)
- The Linux multilib toolchain and the baremetal multilib toolchains have been split into separate tarballs. These now have different expected install paths (`/opt/riscv-linux-toolchain` and `/opt/riscv-baremetal-toolchain`).
- There are new GCC+Clang toolchain tarballs which include both GCC and Clang/LLVM. 
  - These are built by taking the GCC directory and installing clang on top of it, and making a tarball from that "combined" directory.
  - We include `<target>-clang` and `<target>-clang++` executable aliases which are correctly configured for the given target.
- All the toolchain tarballs now use `crosstool-ng` to build GCC (and the sysroot). The Ibex toolchain was already using crosstool-ng, now the multilib toolchains are too.
  - The linux multilib toolchain is not a multilib toolchain due to an issue with crosstool-ng. Once that issue is fixed, we'll have a real multilib toolchain.
- We now generate cross-compilation configuration files for both CMake and Meson, which are included in all the tarballs.

### CODE REVIEW

**Comment by lenary** [comment]
Just looked at the file output of the Ibex build, and the only difference in files in the tar the addition of the cmake and meson configuration files. This is using `tar -t`, so individual file differences are not quantified, but I have a lot of faith in this update. 

I think my next step is to add gcc+clang builds!

**Comment by lenary** [comment]
I'm getting an issue trying to run qemu - `FATAL: kernel too old`. I'll need to investigate where the version mismatch is coming from.

**Comment by lenary** [comment]
TODO:
- [ ] Check Qemu
- [ ] Add `-mcmodel` to toolchain args.

**Comment by lenary** [comment]
I *think* I've managed to make this as relocatable as I'll manage, partly by relying a lot less on the clang config file to pass arguments to Clang, and just putting the paths into the cmake/meson configuration files (which users can edit).

File: build-clang-with-args.sh [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  -DLLVM_DEFAULT_TARGET_TRIPLE="${toolchain_target}" \
+  -DCLANG_VENDOR="lowRISC" \
+  -DBUG_REPORT_URL="toolchains@lowrisc.org" \
+  -DLLVM_INCLUDE_EXAMPLES=Off \
+  -DLLVM_INCLUDE_DOCS=Off \
+  -DLLVM_INSTALL_TOOLCHAIN_ONLY=On \
+  -DLLVM_INSTALL_BINUTILS_SYMLINKS=Off \
+  -DCLANG_LINKS_TO_CREATE="${clang_links_to_create}" \
+  -DLLD_SYMLINKS_TO_CREATE="${lld_links_to_create}" \
+  -DLLVM_DISTRIBUTION_COMPONENTS="clang;clang-format;clang-tidy;clang-resource-headers;lld"
```
Comment (lenary) [inline build-clang-with-args.sh]: - [ ] LLVM's binutils. (objdump, etc)

File: generate-clang-meson-cross-file.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+# to point to the new location of the toolchain.
+
+[binaries]
+c = '${toolchain_dest}/bin/${toolchain_target}-clang'
+cpp = '${toolchain_dest}/bin/${toolchain_target}-clang++'
+ar = '${toolchain_dest}/bin/${toolchain_target}-ar'
+ld = '${toolchain_dest}/bin/${toolchain_target}-ld'
+objdump = '${toolchain_dest}/bin/${toolchain_target}-objdump'
+objcopy = '${toolchain_dest}/bin/${toolchain_target}-objcopy'
+strip = '${toolchain_dest}/bin/${toolchain_target}-strip'
```
Comment (lenary) [inline generate-clang-meson-cross-file.sh]: - [ ] `llvm-ar`
- [ ] `llvm-objdump`
- [ ] `llvm-objcopy`
- [ ] `llvm-strip`

File: lowrisc-toolchain-gcc-multilib-linux.config [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+# Installation prefix directory; the toolchain will end up in exactly
+# this directory. CT_PREFIX_DIR will be added by the
+# ./build-gcc-with-args.sh script
+
+# Don't save tarballs for re-use (needs writable and pre-created
+# CT_LOCAL_TARBALLS_DIR otherwise)
+# CT_SAVE_TARBALLS is not set
+
+# Having errors building linux (due to sched.h issues)
```
Comment (lenary) [inline lowrisc-toolchain-gcc-multilib-linux.config]: - [ ] These issues might be solved by updating to crosstool-ng HEAD? I think my patch was accepted.

File: generate-clang-config.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
Comment (lenary) [inline generate-clang-config.sh:1]: TIL: you don't even need this, it's totally fine just to create `<triple>-clang` and the target will automatically be inferred from the symlink name.

**Comment by lenary** [comment]
This PR lacks the GCC patches for `B` extension which were added to the GCC-only toolchain by @luismarques recently.

File: generate-clang-meson-cross-file.sh (hunk 2 above)
Comment (lenary) [reply generate-clang-meson-cross-file.sh:67]: In the first instance, I don't think we'll move to these utilities immediately, but we should ensure they are included in the tarball so moving to them later is possible.

File: build-clang-with-args.sh (hunk 1 above)
Comment (lenary) [reply build-clang-with-args.sh:100]: - [ ] `scan-build` would be good to include, maybe?

**Comment by luismarques** [comment]
> I'm getting an issue trying to run qemu - `FATAL: kernel too old`. I'll need to investigate where the version mismatch is coming from.

On morty I solved that issue by upgrading the kernel. On my local machine I'm not seeing issues. This incompatibility might have been related to the ABI changes, but it seems fixed now.

File: build-clang-with-args.sh (hunk 1 above)
Comment (luismarques) [reply build-clang-with-args.sh]: - [ ] llvm-readelf
`llvm-readelf` should be a symlink to `llvm-readobj`, as it's not longer an llvm-tool by itself. I need to check how that's currently done.

**Pull request merged**